File: src/main/java/dev/reyaan/wherearemytms/fabric/block/TMMachineScreen.java
package dev.reyaan.wherearemytms.fabric.block;

import com.cobblemon.mod.common.api.moves.Move;
import com.cobblemon.mod.common.api.moves.MoveSet;
import com.cobblemon.mod.common.client.CobblemonClient;
import com.cobblemon.mod.common.pokemon.Gender;
import com.cobblemon.mod.common.pokemon.Pokemon;
import com.cobblemon.mod.common.pokemon.Species;
import com.cobblemon.mod.common.pokemon.activestate.PokemonState;
import com.mojang.blaze3d.systems.RenderSystem;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.minecraft.client.gui.DrawableHelper;
import net.minecraft.client.gui.screen.ingame.HandledScreen;
import net.minecraft.client.render.GameRenderer;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.text.MutableText;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import net.minecraft.util.Identifier;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import static com.cobblemon.mod.common.api.gui.GuiUtilsKt.drawPortraitPokemon;
import static com.cobblemon.mod.common.client.render.RenderHelperKt.drawScaledText;
import static com.cobblemon.mod.common.util.LocalizationUtilsKt.lang;
import static dev.reyaan.wherearemytms.fabric.WhereAreMyTMs.TM_MACHINE_CLOSE_PACKET_ID;
import static dev.reyaan.wherearemytms.fabric.WhereAreMyTMs.id;
import static dev.reyaan.wherearemytms.fabric.block.RenderUtils.s_blitk;

@Environment(EnvType.CLIENT)
public class TMMachineScreen extends HandledScreen<TMMachineScreenHandler> {
    // this code is so cursed lmao
    // i dont even want to fix it
    // rendering is for losers

    // im literally rendering a part of it in this screen, some widgets elsewhere, theyre all hardcoded
    // i have random values, its not symmetrical
    // the buttons are massive and also tiny at the same time
    // theres like 18 for loops
    // for some reason ive done the screen and screenhandler this way instead of pure client side
    // and im STILL USING PACKETS

    public static final Identifier battleInfoBase = id("textures/gui/battle_info_base.png");
    public static final Identifier battleInfoUnderlay = id("textures/gui/battle_info_underlay.png");

    public static final Identifier DEFAULT_LARGE = new Identifier("uniform");

    private static final int INFO_OFFSET_X = 7;

    private static final int TILE_WIDTH = 131;
    private static final int TILE_HEIGHT = 40;
    private static final int PORTRAIT_DIAMETER = 28;
    private static final int PORTRAIT_OFFSET_X = 5;
    private static final int PORTRAIT_OFFSET_Y = 8;

    List<MoveOptionWidget> optionWidgets = new ArrayList<>();
    List<NextButtonWidget> buttonWidgets = new ArrayList<>();

    int currentSlot = -1;
    List<Pokemon> party;

    public TMMachineScreen(TMMachineScreenHandler handler, PlayerInventory inventory, Text title) {
        super(handler, inventory, title);
    }

    @Override
    protected void drawBackground(MatrixStack matrices, float delta, int mouseX, int mouseY) {
        RenderSystem.setShader(GameRenderer::getPositionTexShader);
        RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
        int x = (width - backgroundWidth) / 2;
        int y = (height - backgroundHeight) / 2;
        drawTexture(matrices, x, y, 0, 0, backgroundWidth, backgroundHeight);
    }

    @Override
    public void render(MatrixStack matrices, int mouseX, int mouseY, float delta) {
        this.optionWidgets.clear();

        this.party = CobblemonClient.INSTANCE.getStorage().getMyParty().getSlots();
        if (this.currentSlot == -1) {
            this.currentSlot = getDefaultSlot();
        }

        if (this.currentSlot != -1) {
            Pokemon pokemon = party.get(currentSlot);
            if (pokemon != null) {
                drawMainPokemon(
                        matrices,
                        width / 2.0F - (TILE_WIDTH / 2.0F) - 12,
                        height / 2.0F - (TILE_HEIGHT * 2F),
                        pokemon.getSpecies(),
                        pokemon.getLevel(),
                        pokemon.getAspects(),
                        pokemon.getDisplayName(),
                        pokemon.getGender(),
                        pokemon.getState()
                );

                drawMoveSet(
                        matrices,
                        width / 2.0F - 12,
                        height / 2.0F,
                        pokemon.getMoveSet(),
                        mouseX,
                        mouseY
                );

                drawButtons(
                        matrices,
                        width / 2.0F,
                        height / 2.0F - (TILE_HEIGHT * 1.5F),
                        mouseX,
                        mouseY
                );
            }
        }
    }

    @Override
    protected void init() {
        super.init();
        titleX = (backgroundWidth - textRenderer.getWidth(title)) / 2;
    }


    private void drawMoveSet(
            MatrixStack matrices,
            float originX,
            float originY,
            MoveSet moveSet,
            int mouseX,
            int mouseY
    ) {
        // i hate x and y more than z
        int column = 0;
        int row = 0;

        int xOffset = MoveOptionWidget.TILE_WIDTH;
        int yOffset = MoveOptionWidget.TILE_HEIGHT+12;

        for (Move move : moveSet) {
            float x = originX - (xOffset/2.0F) + (column * (xOffset + 12.0F));
            float y = originY + (row * yOffset);
            MoveOptionWidget widget = new MoveOptionWidget(move.getTemplate(), x, y);
            widget.hovered = widget.isHovered(mouseX, mouseY);
            widget.render(matrices);
            this.optionWidgets.add(widget);

            if (column == 1) {
                row++;
                column = 0;
            } else {
                column++;
            }
        }
    }

    private void drawButtons(
            MatrixStack matrices,
            float originX,
            float originY,
            int mouseX,
            int mouseY
    ) {
        NextButtonWidget back = new NextButtonWidget(true, originX-116, originY);
        NextButtonWidget forward = new NextButtonWidget(false, originX+104, originY);
        buttonWidgets.add(back);
        buttonWidgets.add(forward);

        for (NextButtonWidget widget : buttonWidgets) {
            widget.hovered = widget.isHovered(mouseX, mouseY);
            widget.render(matrices);
        }
    }

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        for (MoveOptionWidget widget : optionWidgets) {
            if (widget.isHovered(mouseX, mouseY)) {
                PacketByteBuf buf = PacketByteBufs.create();
                NbtCompound nbtCompound = new NbtCompound();
                widget.moveTemplate.create().saveToNBT(nbtCompound);
                buf.writeNbt(nbtCompound);
                ClientPlayNetworking.send(TM_MACHINE_CLOSE_PACKET_ID, buf);
                return true;
            }
        }
        // cursed
        // not even gonna bother caching the mons
        for (NextButtonWidget widget : buttonWidgets) {
            if (widget.isHovered(mouseX, mouseY)) {
                if (!widget.back) {
                    for (int i = currentSlot+1; i < party.size(); i++) {
                        Pokemon pokemon = party.get(i);
                        if (pokemon != null) {
                            this.currentSlot = i;
                            return true;
                        }
                    }
                } else if (this.currentSlot != 0) {
                    for (int i = currentSlot-1; i > -party.size()+currentSlot; i--) {
                        int m = i;
                        if (m < 0) {
                            m += party.size();
                        }
                        Pokemon pokemon = party.get(m);
                        if (pokemon != null) {
                            this.currentSlot = m;
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }

    private void drawMainPokemon(
            MatrixStack matrices,
            float x,
            float y,
            Species species,
            int level,
            Set<String> aspects,
            MutableText displayName,
            Gender gender,
            PokemonState state
    ) {
        // Underlay
        float portraitStartX = x + (PORTRAIT_OFFSET_X);
        s_blitk(
                matrices,
                battleInfoUnderlay,
                portraitStartX,
                y + PORTRAIT_OFFSET_Y,
                PORTRAIT_DIAMETER,
                PORTRAIT_DIAMETER
        );

        // Portrait
        DrawableHelper.enableScissor(
                Math.round(portraitStartX),
                Math.round(y + PORTRAIT_OFFSET_Y),
                Math.round(portraitStartX + PORTRAIT_DIAMETER),
                Math.round(y + PORTRAIT_DIAMETER + PORTRAIT_OFFSET_Y)
        );
        MatrixStack matrixStack = new MatrixStack();
        matrixStack.translate(
                portraitStartX + PORTRAIT_DIAMETER / 2.0,
                (double) y + PORTRAIT_OFFSET_Y - 5.0,
                0.0
        );
        matrixStack.push();
        drawPortraitPokemon(
                species,
                aspects,
                matrixStack,
                18F,
                false,
                null
        );
        matrixStack.pop();
        DrawableHelper.disableScissor();

        // Base
        s_blitk(
                matrices,
                battleInfoBase,
                x,
                y,
                TILE_WIDTH,
                TILE_HEIGHT
        );

        // Display Name
        float infoBoxX = x + PORTRAIT_DIAMETER + PORTRAIT_OFFSET_X + INFO_OFFSET_X;
        drawScaledText(
                matrices,
                DEFAULT_LARGE,
                displayName.formatted(Formatting.BOLD),
                infoBoxX,
                y + 7,
                1F,
                1F,
                Integer.MAX_VALUE,
                0x00FFFFFF,
                false,
                true
        );

        // Gender
        if (gender != Gender.GENDERLESS) {
            boolean isMale = gender == Gender.MALE;
            drawScaledText(
                    matrices,
                    DEFAULT_LARGE,
                    Text.literal(isMale ? "♂" : "♀").formatted(Formatting.BOLD),
                    infoBoxX + 53,
                    y + 7,
                    1F,
                    1F,
                    Integer.MAX_VALUE,
                    isMale ? 0x32CBFF : 0xFC5454,
                    false,
                    true
            );
        }

        // Lv. Text
        drawScaledText(
                matrices,
                DEFAULT_LARGE,
                lang("ui.lv").formatted(Formatting.BOLD),
                infoBoxX + 59,
                y + 7,
                1F,
                1F,
                Integer.MAX_VALUE,
                0x00FFFFFF,
                false,
                true
        );

        // Level Number
        drawScaledText(
                matrices,
                DEFAULT_LARGE,
                Text.literal(String.valueOf(level)).formatted(Formatting.BOLD),
                infoBoxX + 72,
                y + 7,
                1F,
                1F,
                Integer.MAX_VALUE,
                0x00FFFFFF,
                false,
                true
        );
    }


    private int getDefaultSlot() {
        int i = 0;
        for (Pokemon pokemon : this.party) {
            if (pokemon != null) {
                return i;
            }
            i++;
        }
        return -1;
    }
}

File: src/main/kotlin/dev/reyaan/wherearemytms/fabric/client/screen/TMMachineScreen.kt
package dev.reyaan.wherearemytms.fabric.client.screen

import com.cobblemon.mod.common.api.gui.blitk
import com.cobblemon.mod.common.api.moves.Move
import com.cobblemon.mod.common.api.text.bold
import com.cobblemon.mod.common.api.text.text
import com.cobblemon.mod.common.client.gui.summary.widgets.ModelWidget
import com.cobblemon.mod.common.client.render.drawScaledText
import com.cobblemon.mod.common.pokemon.Gender
import com.cobblemon.mod.common.pokemon.Pokemon
import com.cobblemon.mod.common.util.lang
import dev.reyaan.wherearemytms.fabric.WAMT.TM_MACHINE_MOVE_SELECT_PACKET_ID
import dev.reyaan.wherearemytms.fabric.WAMT.id
import dev.reyaan.wherearemytms.fabric.client.widgets.MoveOptionWidget
import dev.reyaan.wherearemytms.fabric.client.widgets.SlotChangeWidget
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.widget.ButtonWidget
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.nbt.NbtCompound
import net.minecraft.text.Text
import net.minecraft.util.Identifier


@Environment(EnvType.CLIENT)
class TMMachineScreen(private val party: List<Pokemon?>):
    Screen(Text.literal("TM Machine")) {

    companion object {
        const val BASE_WIDTH = 189
        const val BASE_HEIGHT = 136
        private const val PORTRAIT_SIZE = 66

        val MAIN_BACKGROUND: Identifier = id("textures/gui/summary_base.png")
        private val PORTRAIT_BACKGROUND = id("textures/gui/portrait_background.png")

        val DEFAULT_LARGE = Identifier("uniform")
    }

    private var modelWidget: ModelWidget? = null

    private var backButton: SlotChangeWidget? = null
    private var nextButton: SlotChangeWidget? = null

    private val moveOptions: ArrayList<MoveOptionWidget> = ArrayList()

    private fun getNonNullPartyMembers(party: List<Pokemon?>): List<Pokemon> {
        val nonNullParty = ArrayList<Pokemon>()
        party.forEach { if (it != null) nonNullParty.add(it) }
        return nonNullParty
    }

    private var currentSlot = 0;
    private lateinit var nonNullParty: List<Pokemon>

    private fun refresh(x: Int, y: Int) {
        val pokemon = nonNullParty[currentSlot]
        modelWidget?.pokemon = pokemon.asRenderablePokemon()
        addMoves(pokemon, x, y)
    }

    private fun addMoves(pokemon: Pokemon, x: Int, y: Int) {
        moveOptions.clear()
        val tmMoves = pokemon.species.moves.tmMoves

        pokemon.moveSet.forEachIndexed {i, move ->
            moveOptions.add(
                MoveOptionWidget(
                    pX = x + 88,
                    pY = y + 18 + (27 * i),
                    move = move,
                    disabled = move.template !in tmMoves
                ) {
                    useMoveOption(move)
                }
            )
        }
    }

    fun useMoveOption(move: Move) {
        val buf = PacketByteBufs.create()
        val nbtCompound = NbtCompound()
        move.saveToNBT(nbtCompound)
        buf.writeNbt(nbtCompound)
        ClientPlayNetworking.send(TM_MACHINE_MOVE_SELECT_PACKET_ID, buf)
        super.close()
    }



    override fun init() {
        clearChildren()
        super.init()

        nonNullParty = getNonNullPartyMembers(party)

        if (nonNullParty.isEmpty()) {
            MinecraftClient.getInstance().player?.sendMessage(
                Text.of("You should probably get a Pokemon first..")
            )
            super.close()
        } else {

            val x = (width - BASE_WIDTH) / 2
            val y = (height - BASE_HEIGHT) / 2

            backButton = SlotChangeWidget(
                    pX = x + 10,
                    pY = y + 105,
                    resource = "battle_back"
                ) {
                    currentSlot = maxOf(currentSlot - 1, 0)
                    refresh(x, y)
                }


            nextButton = SlotChangeWidget(
                    pX = x + 12,
                    pY = y - 33,
                    resource = "battle_forward"
                ) {
                    currentSlot = minOf(currentSlot + 1, nonNullParty.size - 1)
                    refresh(x, y)
                }


            val pokemon = nonNullParty[currentSlot]

            modelWidget = ModelWidget(
                pX = x + 5,
                pY = y + 24,
                pWidth = PORTRAIT_SIZE,
                pHeight = PORTRAIT_SIZE,
                pokemon = pokemon.asRenderablePokemon(),
                baseScale = 2F,
                rotationY = 325F,
                offsetY = -10.0
            )

            addMoves(pokemon, x, y)
        }
    }

    override fun render(matrices: MatrixStack, mouseX: Int, mouseY: Int, delta: Float) {
        renderBackground(matrices)

        val x = (width - BASE_WIDTH) / 2
        val y = (height - BASE_HEIGHT) / 2

        // portrait background
        blitk(
            matrixStack = matrices,
            texture = PORTRAIT_BACKGROUND,
            x = x + 5,
            y = y + 27,
            width = PORTRAIT_SIZE,
            height = PORTRAIT_SIZE
        )

        // pokemon model
        modelWidget?.render(matrices, mouseX, mouseY, delta)

        // actual background
        blitk(
            matrixStack = matrices,
            texture = MAIN_BACKGROUND,
            x = x,
            y = y,
            width = BASE_WIDTH,
            height = BASE_HEIGHT
        )


        val pokemon = nonNullParty[currentSlot]
        // Draw labels
        drawScaledText(
            matrixStack = matrices,
            font = DEFAULT_LARGE,
            text = pokemon.displayName.bold(),
            x = x + 5,
            y = y + 9,
            shadow = true
        )

        val gender = pokemon.gender
        if (gender != Gender.GENDERLESS) {
            val isMale = gender == Gender.MALE
            val textSymbol = if (isMale) "♂".text().bold() else "♀".text().bold()
            drawScaledText(
                matrixStack = matrices,
                font = DEFAULT_LARGE,
                text = textSymbol,
                x = x + 5,
                y = y + 15,
                colour = if (isMale) 0x32CBFF else 0xFC5454,
                shadow = true,
                scale=1.25F
            )
        }

        drawScaledText(
            matrixStack = matrices,
            font = DEFAULT_LARGE,
            text = lang("ui.lv").bold(),
            x = x + 61.5,
            y = y + 7.5,
            shadow = true,
            scale=0.6F
        )

        drawScaledText(
            matrixStack = matrices,
            font = DEFAULT_LARGE,
            text = pokemon.level.toString().text().bold(),
            x = x + 61.5,
            y = y + 10.5,
            shadow = true
        )

        moveOptions.forEach {
            it.render(matrices, mouseX, mouseY, delta)
        }

        backButton?.render(matrices, mouseX, mouseY, delta)
        nextButton?.render(matrices, mouseX, mouseY, delta)
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        backButton?.let {
            if (it.isHovered(mouseX, mouseY)) {
                it.onPress()
                return true
            }
        }

        nextButton?.let {
            if (it.isHovered(mouseX, mouseY)) {
                it.onPress()
                return true
            }
        }

        moveOptions.forEach {
            if (it.isHovered(mouseX, mouseY)) {
                it.onPress()
                return true
            }
        }
        return false
    }
}

